### PR TITLE
EAR 1117 - Move previous answers computation clientside

### PR DESCRIPTION
**Pull request by tbaggaley**
## What is the context of this PR?

[JIRA Ticket](https://collaborate2.ons.gov.uk/jira/browse/EAR-1117)

Moves `availablePreviousAnswers`, `availableRoutingAnswers`, `availablePipingAnswers`, `availablePipingMetadata` and `availableRoutingDestinations` clientside.

Uses data already available on the client to compute these things in order to reduce load on server when answers list is large, prevent exposing high time complexity queries to clients, avoid unnecessary network usage & apollo client having to check large lists of answers for equality on each page switch.

#### Piping
- Check all piping menus work as before - for answers, metadata and variables and across all page types

#### Routing
- Check routing answer selector and routing destination selector work as before
- Check validation errors still work as before

#### Question validations
- Check min/max/total validations on question pages work as before with regard to selecting metadata and previous answers
- Check validation errors still work as before

#### Publisher
- Check it ain't become broke 

### ~TODO~ Change summary

- [x] Preliminary profiling of existing code given large questionnaires
- [x] Delete `sharedErrors` resolvers on server for min/max validations - redundant & causing potential DOS issue when many errors present
- [x] Add function to UI to calculate previous answers - `utils/getContentBeforeEntity`
- [x] Calculate previous routing answers clientside
- [x] Calculate previous answers for calculated summary page clientside
- [x] Calculate `availableRoutingDestinations` clientside
- [x] Calculate available piping metadata clientside
- [x] Calculate previous answers for question page validations clientside
- [x] Update ContentPicker v2 (answer picker) so it works as-is with folders in questionnaire tree
- [x] Delete old ContentPicker `shapeTree` - will no longer be required 🎉
- [x] Add `PageContext` as a supplement to the main `QuestionnaireContext` - gets you just the info for the currently-selected page (via `components/QuestionnaireContext/index.js: usePage()`)
- [x] Add `useCurrentPageId()` as a shorthand way of getting React Router's currently selected entity ID (be it page, confirmation, section, etc.) - available in `components/RouterContext.js`
- [x] Search & destroy graphql references to available.*Answers
- [x] Delete resolvers / tests from API
- [x]  Remove references to previous answers etc. from publisher's GetQuestionnaire query

## What to do after everything is green

1. -  Bring this branch up-to-date with Origin/Master
2. - If there are a lot of commits or some are not helpful to read, squash them down
3. - Click the **Merge** button on this pull request
4. - Does this change mean the **Capability examples** questionnaire should be updated?
5. - Move the Jira ticket for this task into the next stage of the process

### CODE REVIEW

File: eq-author/src/App/page/Design/CalculatedSummaryPageEditor/AnswerSelector.js [hunk 1]
```diff
@@ -248,87 +218,94 @@ export class UnwrappedAnswerSelector extends Component {
...
+  };
 
+  const renderEmptyState = (availableSummaryAnswers) => {
     const errorValidationMsg = getValidationError({
       field: "summaryAnswers",
       message: CALCSUM_ANSWER_NOT_SELECTED,
     });
 
-    const title = isAvailableAnswers
+    const title = availableSummaryAnswers.length
```
Comment (tjamesmac) [inline eq-author/src/App/page/Design/CalculatedSummaryPageEditor/AnswerSelector.js:251]:
```suggestion
    const [title, text] = availableSummaryAnswers.length  ? ["No answers selected", "Select an answer using the button below."] : ["No answers available", "There are no answers to provide a calculated summary."]
```

Comment (tjamesmac) [reply eq-author/src/App/page/Design/CalculatedSummaryPageEditor/AnswerSelector.js:251]: i have no idea if that's any better or not, but it's an option

File: eq-author/src/App/page/Design/Validation/PreviousAnswerContentPicker.test.js [hunk 2]
```diff
@@ -1,52 +1,48 @@
...
+}));
+
+jest.mock("components/RouterContext", () => ({
+  useCurrentPageId: jest.fn(),
+}));
+
+useQuestionnaire.mockImplementation(() => ({
+  questionnaire: mockQuestionnaire,
+}));
+useCurrentPageId.mockImplementation(() => "1.1.2");
```
Comment (tjamesmac) [inline eq-author/src/App/page/Design/Validation/PreviousAnswerContentPicker.test.js:39]: I'm liking mocking things this way

File: eq-author/src/components/QuestionnaireContext/index.test.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import React from "react";
+
+import { PageContextProvider, usePage } from "./";
+import { render, screen } from "tests/utils/rtl";
+const PageContextClient = () => {
```
Comment (tjamesmac) [inline eq-author/src/components/QuestionnaireContext/index.test.js:5]:
```suggestion
import { render, screen } from "tests/utils/rtl";

const PageContextClient = () => {
```

**Review by tjamesmac — COMMENTED** [review]
As far as I can see everything works as expected and none of the comments are required

That's some major deletion going on. Nice one!

File: eq-author/src/App/page/Design/CalculatedSummaryPageEditor/AnswerSelector.js (hunk 1 above)
Comment (tbaggaley) [reply eq-author/src/App/page/Design/CalculatedSummaryPageEditor/AnswerSelector.js:251]: Yeah the repetition did bug me a bit too. Tried your thing but post-prettierification it doesn't look very nice. I've moved these variables inline into the JSX instead

File: eq-author/src/components/QuestionnaireContext/index.test.js (hunk 3 above)
Comment (tbaggaley) [reply eq-author/src/components/QuestionnaireContext/index.test.js]: Done, cheers

File: eq-author/src/App/page/Design/Validation/PreviousAnswerContentPicker.test.js (hunk 2 above)
Comment (tbaggaley) [reply eq-author/src/App/page/Design/Validation/PreviousAnswerContentPicker.test.js:39]: Go team!

**Review by tjamesmac — APPROVED** [review]
Niiiiiiiiiiiiiiiiiice

**Comment by tbaggaley** [comment]
NB: Codecov is broken after last merge w/ master - it was fine before though honest! ;-)

Should be +0.06% rather than -2.34%

**Review by farres1 — APPROVED** [review]

**Pull request merged**